src/zap/zapper.cpp

```diff
@@ -1201,6 +1201,8 @@ void Zapper::InitializeCompilerFlags(CORCOMPILE_VERSION_INFO * pVersionInfo)
 
 #endif // _TARGET_X86_
 
+    m_pOpt->m_compilerFlags.Set(CORJIT_FLAGS::CORJIT_FLAG_FEATURE_SIMD);
```

Comment:
`Compiler::compSupportsHWIntrinsic` in RyuJIT would refuse generating SSE/SSE2 without this flag.

We still block loading `Vector<T>` on the runtime side, so vectors won't light up with this.

Reply:
CC. @CarolEidt.

Pretty sure this is safe, but she may be able to better answer if there are any other fixups needed.

Reply:
Do we also need to set the hwintrinsic feature and `EnableSSE`/`EnableSSE2`?

Do we have a test showing that `EnableSSE3+` and others like `EnableLZCNT` or `EnableBMI1` (which aren't part of the SSE hierarchy) don't get generated for crossgen/r2r?

Reply:
There were some wrinkles with enabling the recognition of SIMD types in crossgen (which was needed to correctly support the Vector ABI on Arm64), but I think that we should be OK now that those have been ironed out.

> Do we also need to set the hwintrinsic feature and EnableSSE/EnableSSE2?

The latter two should always be enabled by default. I'm not sure about the hw intrinsic feature.

> Do we have a test showing that EnableSSE3+ and others like EnableLZCNT or EnableBMI1 (which aren't part of the SSE hierarchy) don't get generated for crossgen/r2r?

I don't think that we currently have a good way to detect this. It would certainly worth making an investment in figuring out how we can do this.

Reply:
> The latter two should always be enabled by default. I'm not sure about the hw intrinsic feature.

I guess the better question is there anything we need to do here to ensure that the CPUID checks and the enabled ISAs are correct?

For the normal JIT codegen we ultimately have:

The CPUID checks/caching that happens in the VM: https://github.com/dotnet/coreclr/blob/master/src/vm/codeman.cpp#L1374
-and -
Various JitConfig checks that tell the HWIntrinsic code which ISAs are "supported": https://github.com/dotnet/coreclr/blob/master/src/jit/compiler.cpp#L2254 (since users can disable HWIntrinsic codegen for SSE/SSE2 even if they are "baseline").

> I don't think that we currently have a good way to detect this. It would certainly worth making an investment in figuring out how we can do this.

Could we just have a set of tests that are always run through crossgen/r2r? We could then just hardcode them to validate `Sse.IsSupported`/`Sse2.IsSupported` are true and all other ISAs are false.

Reply:
> The latter two should always be enabled by default. I'm not sure about the hw intrinsic feature.

This one defaults to true as well (they're opt-out). I guess after this change, we won't be able to opt-out in CoreLib (since the code was already generated), but whoever really needs that can opt out of ready to run.

> Could we just have a set of tests that are always run through crossgen/r2r?

I'll need to think about it some more. We currently only enable generating the HW intrinsics for CoreLib, because we know CoreLib won't do something dumb, like caching Avx.IsSupported into a field and using that thorough. It would probably also be weird for end users to step through their code in the debugger and see that we're not taking code paths we're supposed to take (since we're making crossgen more of a user-facing feature now). Limiting weirdness to CoreLib is better.

Reply:
> because we know CoreLib won't do something dumb, like caching Avx.IsSupported into a field and using that thorough

Even if some user did this, I don't think there is a concern. It just means they will get worse codegen at runtime (e.g. they won't hit any Avx codepaths using their cached check).

We should also be actively discouraging people from doing that because it will produce worse codegen (since it may not be treated as a JIT time constant).

I think the perf benefits of enabling this everywhere probably outweighs the chance that someone does something dumb, especially once we start using the HWIntrinsics more broadly (and we already are for some CoreFX code).

Reply:
(I imagine we will also ultimately have an AOT story where SSE/SSE2 are enabled by default and where you can explicitly opt-into additional ISAs if you want to be more restrictive, in which case we probably just want to do the right thing by default here as well)

Reply:
I'd like to see us address #17568 and #21603 more broadly (post 3.0, presumably) - i.e.
- recognize SIMD types during crossgen so that, even if we're not recognizing intrinsics, we don't generate really awful code to reference the structs.
- support intrinsics more broadly during AOT, as @tannergooding suggests above
  - I think that if we allow the developer to opt into additional ISAs, we'll need to come up with a plan for both how to best expose it, as well as how to ensure adequate test coverage.

For now, this seems like a very good initial step to address the startup issues.

Reply:
After this change would there be more work in the runtime to allow crossgen to specify a minimum ISA?  I'm assuming testing, but beyond that nothing fundamental, correct?

Also for those who opt out of tiered jitting, does this make things worse in any way?

Reply:
If one opts out of tiered jitting, then any pessimization caused by SSE/AVX transition penalties would remain. It is also the case that other pessimizations due to R2R constraints would remain.